### ISSUE TITLE
Thoughts about CSS handling

### ISSUE THREAD

**Issue by menelaos** [alexmingoia/purescript-pux#27]
This is supposed to be a discussion thread about CSS in `Pux`.

Currently - due to its use of `react` - `Pux`'s [`style`](http://www.alexmingoia.com/purescript-pux/docs/API/Pux/Html/Attributes.html#style) function takes a record of camelCased versions of CSS properties.

``` purescript
myStyle = { backgroundColor : "red" }
```

As it stands, there are a few things that I think could be improved:
## Type safety

Currently, errors like

``` purescript
myStyle = { backgroundColor : "re" }
```

or

``` purescript
myStyle = { backgroundColo : "red" }
```

are not caught at compile time. By using a solution like [purescript-css](https://pursuit.purescript.org/packages/purescript-css/), this could be remedied.
## Merging CSS

In PureScript, it's not (yet) possible to [combine](http://stackoverflow.com/questions/35985693/combine-records-in-purescript/) two records into one.

As far as I can tell, things like

``` javascript
const myStyleWithRedBorder =
  { ...myStyle
  , border : "1px solid red"
  }
```

are not possible right now in `Pux`.

For this issue, one could adopt `Elm`'s approach of [`style`](http://package.elm-lang.org/packages/evancz/elm-html/4.0.2/Html-Attributes#style) taking an Array of (key, value) pairs.

@alexmingoia I'm sure you have already thought about these issues and so I'd be curious to know your stance on this.

**Comment by alexmingoia** [comment]
I'm glad you brought this up. CSS can and should be type safe and easily composable. 

There are different approaches for type-safe CSS, so I didn't want Pux to dictate how it should be done. As you point out, `purescript-css` can be used to build type-safe (and composable) CSS. One way of using this in Pux is rendering to a string, which is inserted as a `style` element in a component's view. This has the advantage of using media queries, classes, and pseudo-classes which inline styles don't support, and it still keeps CSS specific to a component. I personally think this is the better approach because ultimately you'll need `:hover`.

Using type-safe inline styles is more tricky. React requires inline styles to be specified using an object, so to use purescript-css a rendering method would need to be written that took the CSS and rendered it to an object instead of a string. The problem is that there's no way to write such a render function, because as you pointed out, records can't be merged.

The `style` attribute could be changed to accept an array of key/vals as Elm does, and Pux could transform the array into the object React wants inside the foreign function. Then if someone wants to use inline styles they could merge them, and if they wanted type-safety and composability they could render purescript-css to an array of key/vals.

How does that sound?

**Comment by menelaos** [comment]
Thanks for the detailed response.
Keeping CSS handling separate from `Pux` is definitely a good idea, given that even in the `react` world, there's no consensus yet on how it should be done.

I guess the proposal in your last paragraph also makes sense in that it is the closest to standard CSS.

**Comment by maddie927** [comment]
I agree that it's best to keep the options open, but we should document the known options in the guides.  There's other beginner material I want to contribute soon too, so I might put that all together into another issue's checklist.

**Comment by alexmingoia** [comment]
I've released Pux 4.0.0 which changes `style` to accept an array of tuples. I've also released [pux-css](https://github.com/alexmingoia/pux-css) which provides a `style` attribute that accepts `CSS` from purescript-css. Also added a [CSS section](https://alexmingoia.github.io/purescript-pux/docs/css.html) to the guide.
